psi4/driver/driver.py

```diff
@@ -1932,35 +1932,164 @@ def gdma(wfn, datafile=""):
     Gaussian FCHK format.
 
     .. versionadded:: 0.6
 
     :returns: None
 
+    :type wfn: :py:class:`~psi4.core.Wavefunction`
+    :param wfn: set of molecule, basis, orbitals from which to generate fchk file
+
     :type filename: string
     :param filename: destination file name for FCHK file
 
-    :type wfn: :py:class:`~psi4.core.Wavefunction`
-    :param wfn: set of molecule, basis, orbitals from which to generate fchk file
+    :type debug: boolean
+    :param debug: returns a dictionary to aid with debugging
 
     Notes
     -----
     * A description of the FCHK format is http://wild.life.nctu.edu.tw/~jsyu/compchem/g09/g09ur/f_formchk.htm
     * The allowed headers for methods are general and limited, i.e., "Total SCF|MP2|CI|CC Density",
-      so "CC" is always used for the post-HF case.
+      PSI4 will try to find the right one for the current calculation.
+    * Not all theory modules in PSI4 are compatible with the FCHK writer. Currently available and tested are:
+      All SCF/DFT variants (if gradient avaiable), DF-RHF-MP2 (dfmp2 module), CONV-RHF/UHF-CC methods (ccenergy module), orbital-relaxed DCT (ODC-* functionals) (dct module). No UHF-MP2 densities are available.
+      A warning will be printed if a theory module is not supported.
 
     :examples:
 
     >>> # [1] FCHK file for DFT calculation
     >>> E, wfn = energy('b3lyp', return_wfn=True)
     >>> fchk(wfn, 'mycalc.fchk')
 
+    >>> # [2] FCHK file for correlated densities
+    >>> E, wfn = gradient('ccsd', return_wfn=True)
+    >>> fchk(wfn, 'mycalc.fchk')
+
     """
+    #NOTE: This function should be overhauled once all modules set Ca/Cb, Da/Db in the same manner.
+
+    # QC_MODULE overview for FCHK
+    # OCC: (includes both occ/dfocc modules
+    #  wfn.D() = SCF
+    #  wfn.C() = SCF
+    #  refwfn.D() = SCF
+    # -> not allowed, graceful exit
+    # DCT:
+    #  wfn.D() = SCF
+    #  wnf.C() = DCT
+    #  refwfn.D() = SCF
+    # -> allowed, but FCHK entries are switched around
+    # CCENERGY
+    #  wfn.D() = CC
+    #  wfn.C() = SCF 
+    #  refwfn.D() = CC
+    # -> allowed, will recompute SCF density from wfn.C()
+    # DFMP2:
+    #  wfn.D() = MP2
+    #  wnf.C() = NO(MP2)
+    #  refwfn.D() = SCF 
+    # -> allowed, will take refwfn.D() for SCF density
+
+    allowed = ['DFMP2','SCF','CCENERGY','DCT']
+    module_ = wfn.module().upper()
+    if (module_ not in allowed):
+        core.print_out(f"Theory module {module_} is currently not supported by the FCHK writer.")
+        return None
+
     fw = core.FCHKWriter(wfn)
+
+    # handle special cases
+    if (module_ == 'CCENERGY'):
+        fw.recompute_scf_density(True)
+    elif (module_ == 'DCT'):
+        # tasks:
+        # i)   compute DCT density from wfn.C coefficients and write result to wfn.D
+        # ii)  overwrite wfn.C with SCF orbitals from reference wfn.
+        # iii) let FCHKWriter recompute SCF density
+        wfn_ = wfn
```

Comment:
In Python, this doesn't make `wfn_` a _copy_ of `wfn`. This makes `wfn_` another name for the same object. So when you mutate `wfn_`, you are changing the input `wfn`. Which is bad. I think you want the cloning method for the Wavefunction.

Reply:
right, I had a feeling this isn't right in python. I meant to clone and preserve the original `wfn`.